packages/server/socketHelpers/sendEncodedMessage.ts

```diff
@@ -1,16 +1,57 @@
 import {HttpResponse, WebSocket} from 'uWebSockets.js'
+import handleDisconnect from '../socketHandlers/handleDisconnect'
 import sendSSEMessage from '../sse/sendSSEMessage'
+import ConnectionContext from './ConnectionContext'
 import isHttpResponse from './isHttpResponse'
 
 const ESTIMATED_MTU = 1400
-const sendEncodedMessage = (transport: WebSocket | HttpResponse, message: object | string) => {
-  if (transport.done) return
-  const str = JSON.stringify(message)
-  if (isHttpResponse(transport)) {
-    sendSSEMessage(transport as HttpResponse, str)
+const TIMEOUT_COEFFICIENT = 2.0
+const INITIAL_TIMEOUT = 2000 // 2 seconds
+const MAX_TIMEOUT = 10000 // 10 seconds
+const TIMEOUT_JITTER = 20 // 20 ms
+
+const sendAndPushToReliableQueue = (
+  context: ConnectionContext,
+  mid: number,
+  object: any,
+  timeout = INITIAL_TIMEOUT + Math.random() * TIMEOUT_JITTER
+) => {
+  const {socket, reliableQueue} = context
+  if (timeout > MAX_TIMEOUT) {
+    handleDisconnect(context)
     return
   }
-  transport.send(str, false, str.length > ESTIMATED_MTU)
+  const message = JSON.stringify([object, mid])
+  sendEncodedMessageBasedOnSocket(socket, message)
+  const timer = setTimeout(() => {
+    const newTimeout = timeout * TIMEOUT_COEFFICIENT + Math.random() * TIMEOUT_JITTER
+    sendAndPushToReliableQueue(context, mid, object, newTimeout)
+  }, timeout)
+
+  if (mid in reliableQueue) {
```

Comment:
itty bitty super nitty:
- try to avoid using `in`. it's dangerous. in your browser console, type `'toString' in {}`. notice that it returns true. that's because `in` refers to all methods, not just the enumerable ones. for that reason, you have to follow it with `foo.hasOwnProperty('toString')`, unless `foo` was instantiated with `new Object(null)`, in which case you have to write `Object.prototype.call(foo, 'toString')` :vomiting_face: 
- as an alternative, you can just write `message = reliableQueue[mid]`. the property access cleaner & actually _faster_ than checking if the property exists!